src/libopensc/pkcs15-esteid2018.c

```diff
@@ -0,0 +1,226 @@
+
+    memset(&cert_info, 0, sizeof(cert_info));
+    memset(&cert_obj, 0, sizeof(cert_obj));
+
+    strlcpy(cert_obj.label, esteid_cert_names[i], sizeof(cert_obj.label));
+    sc_format_path(esteid_cert_paths[i], &cert_info.path);
+    cert_info.id.value[0] = esteid_cert_ids[i];
+    cert_info.id.len = 1;
+    r = sc_pkcs15emu_add_x509_cert(p15card, &cert_obj, &cert_info);
+    if (r < 0)
+      LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+    // Read data from first cert
+    if (i != 0)
+      continue;
+
+    sc_pkcs15_cert_t *cert = NULL;
+    r = sc_pkcs15_read_certificate(p15card, &cert_info, &cert);
+    LOG_TEST_RET(card->ctx, r, "Could not read authentication certificate");
+
+    if (cert->key->algorithm == SC_ALGORITHM_EC)
+      field_length = cert->key->u.ec.params.field_length;
+
+    static const struct sc_object_id cn_oid = {{2, 5, 4, 3, -1}};
+    u8 *cn_name = NULL;
+    size_t cn_len = 0;
+    sc_pkcs15_get_name_from_dn(card->ctx, cert->subject, cert->subject_len,
+                               &cn_oid, &cn_name, &cn_len);
+    if (cn_len > 0)
+    {
+      char *token_name = malloc(cn_len + 1);
+      if (token_name)
+      {
+        memcpy(token_name, cn_name, cn_len);
+        token_name[cn_len] = '\0';
+        set_string(&p15card->tokeninfo->label, (const char *)token_name);
+        free(token_name);
+      }
+    }
+    free(cn_name);
+    sc_pkcs15_free_certificate(cert);
+  }
+
+  /* add pins */
+  for (i = 0; i < 3; i++)
+  {
+    static const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
+    static const int esteid_pin_min[3] = {4, 5, 8};
+    static const int esteid_pin_ref[3] = {0x01, 0x85, 0x02};
+    static const int esteid_pin_authid[3] = {1, 2, 3};
+    static const char *esteid_pin_path[3] = {"3F00", "3F00ADF2", "3F00"};
+
+    static const int esteid_pin_flags[3] = {SC_PKCS15_PIN_FLAG_NEEDS_PADDING | SC_PKCS15_PIN_FLAG_INITIALIZED,
+                                            SC_PKCS15_PIN_FLAG_NEEDS_PADDING | SC_PKCS15_PIN_FLAG_INITIALIZED | SC_PKCS15_PIN_FLAG_LOCAL,
+                                            SC_PKCS15_PIN_FLAG_NEEDS_PADDING | SC_PKCS15_PIN_FLAG_INITIALIZED | SC_PKCS15_PIN_FLAG_UNBLOCKING_PIN};
+
+    struct sc_pkcs15_auth_info pin_info;
+    struct sc_pkcs15_object pin_obj;
+
+    memset(&pin_info, 0, sizeof(pin_info));
+    memset(&pin_obj, 0, sizeof(pin_obj));
+
+    sc_format_path(esteid_pin_path[i], &pin_info.path);
+    pin_info.auth_id.len = 1;
+    pin_info.auth_id.value[0] = esteid_pin_authid[i];
+    pin_info.auth_type = SC_PKCS15_PIN_AUTH_TYPE_PIN;
+    pin_info.attrs.pin.reference = esteid_pin_ref[i];
+    pin_info.attrs.pin.flags = esteid_pin_flags[i];
+    pin_info.attrs.pin.type = SC_PKCS15_PIN_TYPE_ASCII_NUMERIC;
+    pin_info.attrs.pin.min_length = esteid_pin_min[i];
+    pin_info.attrs.pin.stored_length = 12;
+    pin_info.attrs.pin.max_length = 12;
+    pin_info.attrs.pin.pad_char = 0xFF;
+    pin_info.tries_left = 3;
```

Comment:
Shouldn't tries_left be `-1` if it's unknown?

Reply:
It gets updated after initial adding and 3 is the specification-defined value.